File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner_oo.js
// Source: https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');
    main();
});

function readLine() {
    return inputString[currentLine++];
}


function main() {
    const N_TEST_CASES = readANumber();
    let prisonersChairNumberToWarn = new Array(N_TEST_CASES);

    for (let i = 0; i < prisonersChairNumberToWarn.length; i++) {
        const TEST_CASE = readTestCase();
        const OBJ = new SaveThePrisoner(TEST_CASE);
        prisonersChairNumberToWarn[i] = OBJ.getPrisonerChairNumberToWarn();
    }

    printResults(prisonersChairNumberToWarn);
}

    function readANumber() {
        return +readLine();
    }

    function readTestCase() {
        const TEST_CASE = readLine().split(' ').map(Number);
        return TEST_CASE;
    }

    class SaveThePrisoner {

        #prisoners;
        #sweets;
        #chairNumberToBegin;
        #prisonerChairNumberToWarn;

        constructor(testCase) {
            this.#prisoners = testCase[0];
            this.#sweets = testCase[1];
            this.#chairNumberToBegin = testCase[2];

            this.#prisonerChairNumberToWarn = this.#saveThePrisoner();
        }

            #saveThePrisoner() {
                let prisonerChairNumberToWarn = this.#chairNumberToBegin + (this.#sweets - 1);

                if (prisonerChairNumberToWarn > this.#prisoners) {
                    prisonerChairNumberToWarn %= this.#prisoners;

                    if (prisonerChairNumberToWarn === 0)
                        prisonerChairNumberToWarn = this.#prisoners;
                }

                return prisonerChairNumberToWarn;
            }

        getPrisonerChairNumberToWarn() {
            return this.#prisonerChairNumberToWarn;
        }
    }

    function printResults(array) {
        array.forEach(element => console.log(element));
    }

File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner_oo.py
# Source: https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

def main():

    N_TEST_CASES = readANumber()
    prisonersChairNumberToWarn = [None] * N_TEST_CASES

    for i in range(N_TEST_CASES):
        TEST_CASE = readTestCase()
        OBJ = SaveThePrisoner(TEST_CASE)
        prisonersChairNumberToWarn[i] = OBJ.getPrisonerChairNumberToWarn()

    printResults(prisonersChairNumberToWarn)


def readANumber():

    number = int(input())
    return number


def readTestCase():

    TEST_CASE = list(map(int, input().split()))
    return TEST_CASE


class SaveThePrisoner:

    def __init__(self, testCase):

        self.__prisoners = testCase[0]
        self.__sweets = testCase[1]
        self.__chairNumberToBegin = testCase[2]

        self.__prisonerChairNumberToWarn = self.__saveThePrisoner()


    def __saveThePrisoner(self):

        prisonerChairNumberToWarn = self.__chairNumberToBegin + (self.__sweets - 1)

        if prisonerChairNumberToWarn > self.__prisoners:
            prisonerChairNumberToWarn %= self.__prisoners

            if prisonerChairNumberToWarn == 0:
                prisonerChairNumberToWarn = self.__prisoners

        return prisonerChairNumberToWarn


    def getPrisonerChairNumberToWarn(self):

        return self.__prisonerChairNumberToWarn


def printResults(ARRAY):

    [print(element) for element in ARRAY]


if __name__ == "__main__":
    main()
